state-network.md

```diff
@@ -1,142 +1,156 @@
 
-#### Account Trie Data
-
-```
-content_key  := network_id | content_type | trie_path | node_hash
-network_id   := uint16
-content_type := 0x01
-trie_path    := nibbles (TODO)
-node_hash    := hash(content)
+> TODO: consult on best way to define trie paths
 ```
+0x00 | Container(path: Nibbles, node_hash: bytes32, state_root: bytes32)
 
-The SSZ sedes used for encoding/decoding
+content_id := sha256(path | node_hash)
+path       := TODO
 
 ```
-Container(network_id: uint16, content_type: uint8, trie_path: byte_list, node_hash: bytes32)
-```
 
+#### Contract Storage Trie Node
 
-#### Contract Storage Trie Data
+An individual trie node from a contract storage trie.
 
 ```
-content_key  := network_id | content_type | address | trie_path | node_hash
-network_id   := uint16
-content_type := 0x02
-address      := bytes20
-trie_path    := nibbles (TODO)
-node_hash    := hash(content)
+0x01 | Container(address: bytes20, path: Nibbles, node_hash: bytes32, state_root: bytes32)
+
+content_id := sha256(address | path | node_hash)
+path       := TODO
 ```
 
-The SSZ sedes used for encoding/decoding
+
+#### Account Trie Proof
+
+A leaf node from the main account trie and accompanying merkle proof against a recent `Header.state_root`
 
 ```
-Container(network_id: uint16, content_type: uint8, address: bytes20, trie_path: byte_list, node_hash: bytes32)
+0x02 | Container(address: bytes20, state_root: bytes32)
+
+content_id := keccak(address)
 ```
 
 
-#### Contract Bytecode
+#### Contract Storage Trie Proof
+
+A leaf node from a contract storage trie and accompanying merkle proof against the `Account.storage_root`.
 
 ```
-content_key  := network_id | content_type | address | code_hash
-network_id   := uint16
-content_type := 0x03
-address      := bytes20
-code_hash    := keccak(content)
+0x03 | Container(address: bytes20, slot: uint256, state_root: bytes32)
+
+content_id := (keccak(address) + keccak(slot)) % 2**256
 ```
 
-The SSZ sedes used for encoding/decoding
+#### Contract Bytecode
+
+The bytecode for a specific contract as referenced by `Account.code_hash`
 
 ```
-Container(network_id: uint16, content_type: uint8, address: bytes20, code_hash: bytes32)
+0x04 | Container(address: bytes20, code_hash: bytes32)
+
+content_id := sha256(address | code_hash)
 ```
 
-### Radius
 
-Nodes on the network broadcast a `radius` value which is used to advertise how much of the overall trie data a node stores.  `radius` is a 256 bit integer.  We define `MAX_RADIUS = 2**256 - 1`
+### Distance Function
 
-A node is expected to be *interested* in a piece of content if `distance(node_id, content_id) <= radius`.
+The overlay DHT uses the following distance function for determining both:
 
-Nodes are expected to track the radius of other nodes on the network.
+* The distance between two DHT nodes in the network.
+* The distance between a DHT node and a piece of content.
 
+```python
+MODULO = 2**256
+MID = 2**255
 
-### Gossip
+def distance(node_id: int, content_id: int) -> int:
```

Comment:
Just a question, as in most client and discovery v5 implementations we interpret id-s as `Uint256` numbers wouldn't  
```
proc distance(node_id: UInt256, content_id: UInt256): UInt256 =
  let rawDiff = 
    if node_id > content_id:
      node_id - content_id
    else:
      content_id - node_id

  if rawDiff > MID:
    return u256(0) - rawDiff
  else:
    return rawDiff
```
be equivalent to this distance function ? It passes all test cases, and most probably it would be also better performing.

Reply:
If this is equivalent then yes, we can use this more simplified version.  I just pulled something from stack exchange when I came up with the first one to be quick.